src/cpp/session/modules/SessionPackages.R

```diff
@@ -23,531 +23,176 @@
+.rs.addJsonRpcHandler("is_package_attached", function(packageName, libraryPath)
 {
-   reportPackageStatus <- function(status)
-      function(pkgname, ...)
-      {
-         packageStatus = list(name=pkgname,
-                              path=.rs.createAliasedPath(
-                                     .rs.pathPackage(pkgname, quiet=TRUE)),
-                              loaded=status)
-         .rs.enqueClientEvent("package_status_changed", packageStatus)
-      }
+   # quick check first if the package is loaded
+   if (!isNamespaceLoaded(packageName))
+      return(.rs.scalar(FALSE))
    
-   notifyPackageLoaded <- function(pkgname, ...)
-   {
-      .Call("rs_packageLoaded", pkgname, PACKAGE = "(embedding)")
-
-      # when a package is loaded, it can register S3 methods which replace overrides we've
-      # attached manually; take this opportunity to reattach them.
-      .rs.reattachS3Overrides()
-   }
-
-   notifyPackageUnloaded <- function(pkgname, ...)
-   {
-      .rs.beforePackageUnloaded(pkgname)
-      .Call("rs_packageUnloaded", pkgname, PACKAGE = "(embedding)")
-   }
+   # get the raw path to the package
+   packagePath <- .rs.pathPackage(packageName, quiet = TRUE)
+   if (length(packagePath) == 0L)
+      return(.rs.scalar(FALSE))
    
-   pkgNames <-
-      base::list.dirs(.libPaths(), full.names = FALSE, recursive = FALSE)
+   # get associated library path
+   packageLibraryPath <- dirname(packagePath)
    
-   sapply(pkgNames, function(packageName)
-   {
-      if ( !(packageName %in% .rs.hookedPackages) )
-      {
-         attachEventName = packageEvent(packageName, "attach")
-         setHook(attachEventName, reportPackageStatus(TRUE), action="append")
-         
-         loadEventName = packageEvent(packageName, "onLoad")
-         setHook(loadEventName, notifyPackageLoaded, action="append")
-
-         unloadEventName = packageEvent(packageName, "onUnload")
-         setHook(unloadEventName, notifyPackageUnloaded, action="append")
-             
-         detachEventName = packageEvent(packageName, "detach")
-         setHook(detachEventName, reportPackageStatus(FALSE), action="append")
-          
-         .rs.setVar("hookedPackages", append(.rs.hookedPackages, packageName))
-      }
-   })
-})
-
-.rs.addFunction("packages.initialize", function()
-{  
-   # list of packages we have hooked attach/detach for
-   .rs.setVar("hookedPackages", character())
-
-   # set flag indicating we should not ignore loadedPackageUpdates checks
-   .rs.setVar("ignoreNextLoadedPackageCheck", FALSE)
-    
-   # ensure we are subscribed to package attach/detach events
-   .rs.updatePackageEvents()
-})
-
-.rs.addFunction( "addRToolsToPath", function()
-{
-    .Call("rs_addRToolsToPath", PACKAGE = "(embedding)")
+   # compare with the library given by the client
+   samePath <- identical(
+      normalizePath(libraryPath, winslash = "/", mustWork = FALSE),
+      normalizePath(packageLibraryPath, winslash = "/", mustWork = FALSE)
+   )
```

Comment:
This fixes a bug where we could fail to detect if a package was attached. (It appears that `dirname(path)` may also return a normalized version of the directory name; it will not necessarily preserve e.g. a leading tilde for aliased paths.)